File: src/KisanMitraAI.Core/Models/ChartData.cs
namespace KisanMitraAI.Core.Models;

/// <summary>
/// Represents formatted data for charting
/// </summary>
public record ChartData(
    string Title,
    string XAxisLabel,
    string YAxisLabel,
    IEnumerable<ChartSeries> Series,
    IEnumerable<ChartAnnotation> Annotations);

/// <summary>
/// Represents a data series in a chart
/// </summary>
public record ChartSeries(
    string Name,
    IEnumerable<ChartPoint> Points,
    string Color,
    ChartSeriesType Type);

/// <summary>
/// Represents a single point in a chart
/// </summary>
public record ChartPoint(
    string Label,
    decimal Value,
    DateTimeOffset Timestamp);

/// <summary>
/// Represents an annotation on a chart (e.g., anomaly marker)
/// </summary>
public record ChartAnnotation(
    DateTimeOffset Timestamp,
    string Text,
    AnnotationType Type);

/// <summary>
/// Types of chart series
/// </summary>
public enum ChartSeriesType
{
    Line,
    Bar,
    Area,
    Scatter
}

/// <summary>
/// Types of chart annotations
/// </summary>
public enum AnnotationType
{
    Anomaly,
    Peak,
    Trough,
    Milestone
}

/// <summary>
/// Represents comparison chart data for multiple periods
/// </summary>
public record ComparisonChartData(
    string Title,
    IEnumerable<PeriodChartData> Periods,
    IEnumerable<string> Insights);

/// <summary>
/// Represents chart data for a specific period
/// </summary>
public record PeriodChartData(
    string PeriodLabel,
    decimal AverageValue,
    decimal TotalValue,
    int DataPointCount,
    IEnumerable<ChartPoint> Points);

/// <summary>
/// Represents a significant change in the data
/// </summary>
public record SignificantChange(
    DateTimeOffset Timestamp,
    decimal OldValue,
    decimal NewValue,
    decimal ChangePercent,
    string Description);

/// <summary>
/// Represents trend line data for visualization
/// </summary>
public record TrendLineData(
    IEnumerable<ChartPoint> ActualPoints,
    IEnumerable<ChartPoint> TrendLinePoints,
    TrendDirection Direction,
    decimal Slope,
    string Equation);

File: react-frontend/src/types/chartData.ts
// Chart data types matching backend models

export enum ChartSeriesType {
  Line = 'Line',
  Bar = 'Bar',
  Area = 'Area',
  Scatter = 'Scatter'
}

export enum AnnotationType {
  Anomaly = 'Anomaly',
  Peak = 'Peak',
  Trough = 'Trough',
  Milestone = 'Milestone'
}

export enum TrendDirection {
  Increasing = 'Increasing',
  Decreasing = 'Decreasing',
  Stable = 'Stable'
}

export interface ChartPoint {
  label: string;
  value: number;
  timestamp: string;
}

export interface ChartSeries {
  name: string;
  points: ChartPoint[];
  color: string;
  type: ChartSeriesType;
}

export interface ChartAnnotation {
  timestamp: string;
  text: string;
  type: AnnotationType;
}

export interface ChartData {
  title: string;
  xAxisLabel: string;
  yAxisLabel: string;
  series: ChartSeries[];
  annotations: ChartAnnotation[];
}

export interface PeriodChartData {
  periodLabel: string;
  averageValue: number;
  totalValue: number;
  dataPointCount: number;
  points: ChartPoint[];
}

export interface ComparisonChartData {
  title: string;
  periods: PeriodChartData[];
  insights: string[];
}

export interface SignificantChange {
  timestamp: string;
  oldValue: number;
  newValue: number;
  changePercent: number;
  description: string;
}

export interface TrendLineData {
  actualPoints: ChartPoint[];
  trendLinePoints: ChartPoint[];
  direction: TrendDirection;
  slope: number;
  equation: string;
}

export type TimeRange = '7days' | '30days' | 'season' | 'year' | 'custom';

export interface CustomTimeRange {
  startDate: Date;
  endDate: Date;
}

export interface HistoricalDataFilter {
  timeRange: TimeRange;
  customRange?: CustomTimeRange;
  dataType: 'prices' | 'soilHealth' | 'qualityGrades' | 'all';
}

export interface ExportOptions {
  format: 'pdf' | 'csv';
  includeCharts: boolean;
  includeInsights: boolean;
}
